numpy/random/src/mt19937/mt19937-jump.c

```diff
@@ -187,11 +187,9 @@ void mt19937_jump_state(mt19937_state *state, const char *jump_str) {
   pf = (unsigned long *)calloc(P_SIZE, sizeof(unsigned long));
 
   for (i = MEXP - 1; i > -1; i--) {
-    if (jump_str[i] == '1')
+    if (jump_str[(MEXP - 1) - i] == '1')
       set_coef(pf, i, 1);
   }
-  /* TODO: Should generate the next set and start from 0, but doesn't matter ??
```

Comment:
🤔 

In this case I'd maybe call `mt19937_gen` before jumping. As in https://github.com/numpy/numpy/blob/28f9cbc5e5510cd780b9d6a18a8e8bc6b45a13d7/numpy/random/src/mt19937/mt19937.h#L27-L43

Reseting pos to 0 when pos is N means that the final generator after jump will be: 0 + 2\*\*128
While if pos was N-2 it would be: N - 2 + 2\*\*128
While if pos was N-1 it would be: N - 1 + 2\*\*128

Reply:
I think the best we can do here is to have fidelity with the original jump implementation.  Going off in other directions seems unnecessarily risky.

Reply:
I think I agree, except some ease of mind, I honestly do not know why there would be a use-case to generate the same identical stream after different order of operation? The only thing is could be to document it, but since this is the reference implementation...

Reply:
I totally agree that if the problem comes from the reference implementation then it is another issue.

What I was saying is more that given the documentation of this method, as a user I expect that calling jump will offset the generator by `2**128` iterations which should be the same as calling `next` `2**128` times in a row. Saying that `next(jump) !== jump(next)` somehow points that this is not the case 🤔

Reply:
@bashtage maybe we can just tweak the docs a little to be clear that it only jumps about `2**128` states, `2**128` states from the 0 offset postion (whatever it is)?

Reply:
@seberg I added some additional docs explaining that this matches the reference implementation and is tested against it. I'm not totally convinced that the last part of your statement is true in the original code (or this version).  I'm going to try and contact the author to see if he has any notes to share.

Reply:
People still need a mental model of what's going on. From what you've added, I'd still expect `next(jumped()) == jumped(next())` to hold true. If that's not true, then what are the guarantees that we can rely on to reason about our programs?

Reply:
My guess is that the jump is for the whole state and ignores the offset for the next item. @bashtage Is that the case? If so, it should be possible to work around that.

Reply:
>  I'm going to try and contact the author to see if he has any notes to share

@bashtage Did you succeed into getting in touch with the author of the paper? I'd really be interested to know the reason why jump seems to jump at very different locations based on how many times we offset the rng before jumping 🤔 

As a reminder my original issue was:

```py
from numpy.random import MT19937

rng = MT19937(42)
rng.random_raw()   # value given state offset of 0
rng.random_raw()   # value given state offset of 1
rng = rng.jumped()
rng.random_raw()   # value given state offset of 2**128 +2
# Got: 2382575482

rng = MT19937(42)
rng.random_raw()   # value given state offset of 0
rng = rng.jumped()
rng.random_raw()   # value given state offset of 2**128 +1
rng.random_raw()   # value given state offset of 2**128 +2
# Got: 2001582899
```

It seems that the current version of numpy (1.19.4) still has a similiar behaviour. Which seems unexpected given what jump should do 🤔

Reply:
@bashtage Just bringing up the point of *dubzzz* again: Did you hear back anything of the original author of the jump code paper?